src/cgw_ucentral_ap_parser.rs

```diff
@@ -1,48 +1,680 @@
+                warn!("Failed to parse clients port for {:?}, skipping", client);
+                continue;
+            }
+        };
+
+        // Skip wireless clients info
+        if local_port.contains("wlan") || local_port.contains("WLAN") {
+            continue;
+        }
+
+        // TODO: W/A for now: ignore any upstream-reported clients,
+        // because it includes ARP neighbours and clients.
+        // The logic to process uplink neighbors properly should be
+        // much more complicated and should be treated as a
+        // separate case.
+        // This logic also overlaps the uplink switch's neighbor
+        // detection logic (LLDP, ARP, FDB etc), and building topo map
+        // based purely on AP's input could result in invalid
+        // map formation.
+        if upstream_ifaces.iter().any(|i| *i == local_port) {
+            continue;
+        }
+
+        let remote_serial = MacAddress::from_str(client["mac"].as_str().unwrap()).unwrap();
+
+        links.push(CGWUCentralEventStateClients {
+            // Wired clients don't have <connected since> data.
+            // Treat <now> as latest connected ts.
+            client_type: CGWUCentralEventStateClientsType::Wired(timestamp),
+            local_port,
+            remote_serial,
+            // TODO: rework remote_port to have speed / duplex characteristics
+            // for an edge.
+            remote_port: "<Wired-client>".to_string(),
+            is_downstream: true,
+        });
+    }
 }
 
-pub fn cgw_ucentral_ap_parse_message(message: &String) -> Result<CGWUCentralEvent, &'static str> {
+fn parse_interface_data(
+    interface: &Map<String, Value>,
+    links: &mut Vec<CGWUCentralEventStateClients>,
+    upstream_ifaces: &Vec<String>,
+    timestamp: i64,
+) {
+    if interface.contains_key("clients") {
+        if let Value::Array(clients) = &interface["clients"] {
+            parse_wired_clients_data(&clients, links, upstream_ifaces, timestamp);
+        }
+    }
+
+    if interface.contains_key("ssids") {
+        let mut ssids_map: HashMap<String, (String, String)> = HashMap::new();
+        if let Value::Array(ssids) = &interface["ssids"] {
+            parse_wireless_ssids_info(ssids, &mut ssids_map);
+            parse_wireless_clients_data(&ssids, links, upstream_ifaces, &ssids_map, timestamp);
+        }
+    }
+}
+
+fn parse_link_state_data(
+    link_state: &Map<String, Value>,
+    upstream_ifaces: &mut Vec<String>,
+    downstream_ifaces: &mut Vec<String>,
+) {
+    if let Value::Object(upstream_obj) = &link_state["upstream"] {
+        for (k, _v) in upstream_obj.iter() {
+            upstream_ifaces.push(String::from(k.as_str()));
+        }
+    }
+
+    if let Value::Object(downstream_obj) = &link_state["downstream"] {
+        for (k, _v) in downstream_obj.iter() {
+            downstream_ifaces.push(String::from(k.as_str()));
+        }
+    }
+}
+
+fn parse_state_event_data(
+    map: CGWUcentralJRPCMessage,
```

Comment:
May I ask you to rename CGWUcentralJRPCMessage -> CGWUCentralJRPCMessage

To be more consistence with other messages name like "CGWUCentralEvent" etc...